3.Linear_Modeling/nbconverted/fit_linear_model.py

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import pathlib
+import sys
+
+import pandas as pd
+import plotly.express as px
+import seaborn as sns
+from matplotlib import pyplot as plt
+from pycytominer.cyto_utils import infer_cp_features
+from sklearn.linear_model import LinearRegression
+
+sys.path.append("..")
+# from ..utils.utils import df_stats
+import matplotlib.pyplot as plt
+
+# In[2]:
+
+
+def linear_model_function(control, treatment, df, cp_features):
+    print(control[0])
+    print(treatment[0])
+    dosage_treatments_list = treatment + control
+
+    df = df.query("Metadata_Treatment_and_Dose == @dosage_treatments_list")
+    # Add dummy matrix of categorical genotypes
+    treatment_df = pd.get_dummies(data=df.Metadata_Treatment_and_Dose)
+    # Setup linear modeling framework
+    variables = ["Metadata_number_of_singlecells"]
+    X = df.loc[:, variables]
+
+    X = pd.concat([X, treatment_df], axis=1)
+
+    columns_list = []
+    columns_list.append("feature")
+    columns_list.append("r2_score")
+
+    for i in X:
+        columns_list.append(i)
+    # Fit linear model for each feature
+    lm_results = []
+    for cp_feature in cp_features:
+        # Subset CP data to each individual feature (univariate test)
+        cp_subset_df = df.loc[:, cp_feature]
+
+        # Fit linear model
+        lm = LinearRegression(fit_intercept=True)
+        lm_result = lm.fit(X=X, y=cp_subset_df)
+
+        # Extract Beta coefficients
+        # (contribution of feature to X covariates)
+        coef = list(lm_result.coef_)
+        # Estimate fit (R^2)
+        r2_score = lm.score(X=X, y=cp_subset_df)
+
+        # Add results to a growing list
+        lm_results.append([cp_feature, r2_score] + coef)
+
+    # Convert results to a pandas DataFrame
+    lm_results = pd.DataFrame(lm_results, columns=columns_list)
+
+    # Output file
+    output_dir = pathlib.Path("results")
+    output_cp_file = pathlib.Path(
+        output_dir,
+        f"Linear_Model_cp_features_{treatment}_beta_coefficient_against_{control}_beta_coefficient.tsv",
+    )
+    print(output_cp_file)
+    lm_results.to_csv(output_cp_file, sep="\t", index=False)
```

Comment:
Generally, I like to avoid outputting files from functions - it hides important side effects. Instead, consider saving the files separately, in the for loop (as I mentioned previously, I don't think this function is even necessary! 😄 )

Ultimately, this is up to you to decide

Reply:
I also recommend outputting only 1 file (with appropriate column header identifiers). This makes it _way_ easier to wrangle later.